Libplanet.Net.Tests/Consensus/Context/ContextNonProposerTest.cs

```diff
@@ -16,97 +15,99 @@ public class ContextNonProposerTest : ContextTestBase
+            Context.StateChanged += (sender, state) =>
+            {
+                if (state.Round == 1 && state.Step == Step.PreVote)
+                {
+                    stateChangedToRoundOnePreVote.Set();
+                }
+            };
+            Context.StartAsync();
 
-            Context.AddMessage(
+            Context.ProduceMessage(
                 TestUtils.CreateConsensusPropose(
                     block, TestUtils.PrivateKeys[1], round: 0));
 
-            Context.AddMessage(
+            Context.ProduceMessage(
                 TestUtils.CreateConsensusPropose(
                     block, TestUtils.PrivateKeys[2], round: 1));
 
-            AsyncAutoResetEvent messageProcessed = WatchMessageProcessed();
-
-            Context.HandleMessage(
+            Context.ProduceMessage(
                 new ConsensusVote(
                     TestUtils.CreateVote(
                         TestUtils.PrivateKeys[2], 1, 1, hash: block.Hash, flag: VoteFlag.Absent))
                 {
                     Remote = new Peer(TestUtils.Validators[2]),
                 });
 
-            await messageProcessed.WaitAsync();
-            Assert.Equal(Step.Propose, Context.Step);
+            // Wait for round 1 prevote step.
+            await stateChangedToRoundOnePreVote.WaitAsync();
+            Assert.Equal(Step.PreVote, Context.Step);
             Assert.Equal(1, Context.Height);
             Assert.Equal(1, Context.Round);
         }
 
         [Fact(Timeout = Timeout)]
         public async void EnterPreCommitBlockTwoThird()
         {
-            var stepChanged = new AsyncAutoResetEvent();
-            var messageReceived = new AsyncAutoResetEvent();
-            BlockHash? blockHash = null;
-            void IsVoteSent(ConsensusMessage consensusMessage)
+            var block = await BlockChain.MineBlock(TestUtils.PrivateKeys[1], append: false);
+            var blockHash = block.Hash;
+            var stepChangedToPreCommit = new AsyncAutoResetEvent();
+            var commitSent = new AsyncAutoResetEvent();
+            Context.StateChanged += (sender, state) =>
             {
-                if (consensusMessage is ConsensusCommit commit)
+                if (state.Step == Step.PreCommit)
                 {
-                    messageReceived.Set();
-                    Assert.Equal(commit.BlockHash, blockHash);
+                    stepChangedToPreCommit.Set();
                 }
-            }
-
-            watchConsensusMessage = IsVoteSent;
-            Context.StepChanged += (sender, step) =>
+            };
+            ConsensusMessageSent += (observer, message) =>
             {
-                if (step == Step.PreCommit)
+                if (message is ConsensusCommit commit)
                 {
-                    stepChanged.Set();
+                    Assert.Equal(commit.BlockHash, blockHash);
+                    commitSent.Set();
                 }
```

Comment:
I remember of conversation about whether this catches exception internally. Is it okay to have this?

Reply:
Due to ordering, `Assert` coming before `Set()` would break the test as failing `Assert` would result in `Set()` never being run, i.e. event never happening. As for using `Assert`, it is not much different from using `if` statement from unit test framework's perspective.

Reply:
Ah, I have misunderstood with exception being caught internally in `Task`.